### PR TITLE
Added vibration alert enhancement

### PR DESCRIPTION
**Pull request by siva-subramaniam-v**
Feature Request #123 

**Vibrating after receiving files:**

With these changes, the device vibrates every time when a file is received.

The code to vibrate the device is called inside the event listener for the event **file-received** which is fired every time after a file is successfully received. This might not be the ideal solution for transfers involving multiple files as the device vibrates on receiving every single file which the user might find annoying.

**Vibrating after sending files:**

In [**network.js**](https://github.com/RobinLinus/snapdrop/blob/master/client/scripts/network.js) at line 210 the **_onTransferCompleted** function is executed every time after a file is successfully sent. But no such event as **file-received** seems to be fired. I'm unsure how to execute the Android function whenever the **_onTransferCompleted** javascript callback is executed.

### CODE REVIEW

**Comment by siva-subramaniam-v** [comment]
@fm-sys I think it'd be most appropriate to vibrate the device once the entire transfer is complete, rather than vibrating for every file. Let me know if you know any callback/event that is fired when the entire transfer is complete.

**Comment by fm-sys** [comment]
> Let me know if you know of any callback/event that is fired when the entire transfer is complete.

On pairdrop the callback should probably be called only once AFAIK. Can you confirm?

No solution for snapdrop.net though...

File: app/src/main/assets/init.js [hunk 1]
```diff
@@ -173,6 +173,21 @@ try {
 
 window.addEventListener('file-received', e => {
     SnapdropAndroid.saveDownloadFileName(e.detail.name, e.detail.size);
+    // vibrates after receiving each file (SnapDrop and PairDrop)
+    //SnapdropAndroid.vibrate();
+}, false);
+
+window.addEventListener('files-received', e => {
```
Comment (siva-subramaniam-v) [inline app/src/main/assets/init.js:180]: This works as intended but only on PairDrop.

File: app/src/main/assets/init.js [hunk 2]
```diff
@@ -173,6 +173,21 @@ try {
 
 window.addEventListener('file-received', e => {
     SnapdropAndroid.saveDownloadFileName(e.detail.name, e.detail.size);
```
Comment (siva-subramaniam-v) [inline app/src/main/assets/init.js:175]: I wouldn't recommend this as the device vibrates multiple times for a single transfer which is annoying.

File: app/src/main/assets/init.js [hunk 3]
```diff
@@ -173,6 +173,21 @@ try {
...
+window.addEventListener('files-received', e => {
+    // vibrates after receiving all files (supported only on PairDrop)
+    SnapdropAndroid.vibrate();
+}, false);
+
+window.addEventListener('notify-user', e => {
+    if (e.detail == 'File transfer completed.') {
+        // vibrates after sending all files (PairDrop)
+        // vibrates after sending each file (SnapDrop)
+        SnapdropAndroid.vibrate();
```
Comment (siva-subramaniam-v) [inline app/src/main/assets/init.js:189]: Works as intended on PairDrop, but is annoying on SnapDrop.

**Comment by siva-subramaniam-v** [comment]
I've covered all cases and leaving the decision to you.

**Comment by siva-subramaniam-v** [comment]
@fm-sys can you review this?

**Comment by fm-sys** [comment]
I'm so sorry, I'm totally busy currently but I hope that a can have a closer look soon 🙏

File: app/src/main/assets/init.js [hunk 4]
```diff
@@ -173,6 +173,21 @@ try {
 
 window.addEventListener('file-received', e => {
     SnapdropAndroid.saveDownloadFileName(e.detail.name, e.detail.size);
+    // vibrates after receiving each file (SnapDrop and PairDrop)
+    //SnapdropAndroid.vibrate();
+}, false);
+
+window.addEventListener('files-received', e => {
+    // vibrates after receiving all files (supported only on PairDrop)
+    SnapdropAndroid.vibrate();
```
[the remark this answers is not in the document]
Comment (fm-sys) [reply app/src/main/assets/init.js:182]: This would probably be the way to go, IMHO.

File: app/src/main/assets/init.js (hunk 3 above)
Comment (fm-sys) [reply app/src/main/assets/init.js:189]: Seeing that the notification text could change over time, this doesn't seem to be a very robust implementation...

File: app/src/main/assets/init.js [hunk 5]
```diff
@@ -173,6 +173,21 @@ try {
 
 window.addEventListener('file-received', e => {
     SnapdropAndroid.saveDownloadFileName(e.detail.name, e.detail.size);
+    // vibrates after receiving each file (SnapDrop and PairDrop)
+    //SnapdropAndroid.vibrate();
```
[the remark this answers is not in the document]
Comment (fm-sys) [reply app/src/main/assets/init.js:177]: Agreed.

**Review by fm-sys — COMMENTED** [review]
Finally I found time to review the changes...

File: app/src/main/assets/init.js (hunk 3 above)
Comment (fm-sys) [reply app/src/main/assets/init.js:189]: Update: PairDrop now fires a ` 'files-sent' ` event

**Comment by fm-sys** [comment]
Another thing which we should keep in mind:

Some Users might not like vibration effects. Maybe we should either make them optional or disable them automatically if the device is in silence mode. What do you think? 

Helpful links:
https://stackoverflow.com/questions/2048427/how-can-i-detect-whether-the-android-phone-in-silent-mode-programmatically

**Comment by siva-subramaniam-v** [comment]
Agree. The user would expect the app not to vibrate when the device is in silent mode.

We can also add a preference to toggle vibration on or off.

**Comment by fm-sys** [comment]
> We can also add a preference to toggle vibration on or off.

I would probably leave it out for now. Increased UI complexity is IMHO not worth it here...

**Comment by fm-sys** [comment]
Thank you very much!

**Pull request merged**